File: Backend/Core/src/main/scala/APIs/EnrollmentService.scala
package APIs

import Common.API.API
import Global.EnrollmentService
import Objects._

import io.circe.generic.auto._

case class CreateSelectionMessage(
    studentToken: String,
    termId: String,
    subjectId: String,
    courseId: String
) extends API[Unit](EnrollmentService, "CreateSelectionMessage")

case class QueryAllocatedStakePointsMessage(
    studentId: String,
    termId: String
) extends API[Int](EnrollmentService, "QueryAllocatedStakePointsMessage")

case class QuerySelectionsMessage(
    studentId: String,
    termId: String
) extends API[List[Selection]](
      EnrollmentService,
      "QuerySelectionsMessage"
    )

case class UpdateStakeMessage(
    studentToken: String,
    termId: String,
    selectionId: String,
    points: Int
) extends API[Unit](EnrollmentService, "UpdateStakeMessage")

case class UpdateSelectedCourseMessage(
    studentToken: String,
    termId: String,
    selectionId: String,
    courseId: String
) extends API[Unit](EnrollmentService, "UpdateSelectedCourseMessage")

case class DeleteSelectionMessage(
    studentToken: String,
    termId: String,
    selectionId: String
) extends API[Unit](EnrollmentService, "DeleteSelectionMessage")

case class ExecuteEnrollmentMessage(
    adminToken: String,
    termId: String,
) extends API[Unit](EnrollmentService, "ExecuteEnrollmentMessage")

File: Frontend/services/enrollmentService.ts
import { Selection } from "@/types/enrollmentService";
import ApiClient from "./apiClient";

const port = 10013;
const client = new ApiClient(port);

const EnrollmentService = {
  createSelection: client.api<
    {
      studentToken: string;
      termId: string;
      subjectId: string;
      courseId: string;
    },
    void
  >("CreateSelectionMessage"),

  queryAllocatedStakePoints: client.api<
    { studentId: string; termId: string },
    number
  >("QueryAllocatedStakePointsMessage"),

  querySelections: client.api<
    { studentId: string; termId: string },
    Selection[]
  >("QuerySelectionsMessage"),

  updateStake: client.api<
    {
      studentToken: string;
      termId: string;
      selectionId: string;
      points: number;
    },
    void
  >("UpdateStakeMessage"),

  updateSelectedCourse: client.api<
    {
      studentToken: string;
      termId: string;
      selectionId: string;
      courseId: string;
    },
    void
  >("UpdateSelectedCourseMessage"),

  deleteSelection: client.api<
    { studentToken: string; termId: string; selectionId: string },
    void
  >("DeleteSelectionMessage"),

  executeEnrollment: client.api<{ adminToken: string; termId: string }, void>(
    "ExecuteEnrollmentMessage"
  ),
};

export default EnrollmentService;
